firestore/api/SolutionCounter/Program.cs

```diff
@@ -0,0 +1,137 @@
+        /// subcollection of specified document.
+        /// </summary>
+        /// <param name="docRef">The document reference <see cref="DocumentReference"/></param>
+        /// <returns>The <see cref="Task"/></returns>
+        private static async Task CreateCounterAsync(DocumentReference docRef, int numOfShards)
+        {
+            CollectionReference colRef = docRef.Collection("shards");
+            var tasks = new List<Task>();
+            // Initialize each shard with Count=0
+            for (var i = 0; i < numOfShards; i++)
+            {
+                tasks.Add(colRef.Document(i.ToString()).SetAsync(new Shard() { Count = 0 }));
+            }
+            await Task.WhenAll(tasks);
+        }
+        // [END fs_create_counter]
+
+        // [START fs_increment_counter]
+        /// <summary>
+        /// Increment a randomly picked shard by 1.
+        /// </summary>
+        /// <param name="docRef">The document reference <see cref="DocumentReference"/></param>
+        /// <returns>The <see cref="Task"/></returns>
+        private static async Task IncrementCounterAsync(DocumentReference docRef, int numOfShards)
+        {
+            int documentId;
+            lock (s_randLock)
+            {
+                documentId = s_rand.Next(numOfShards);
+            }
+            var shardRef = docRef.Collection("shards").Document(documentId.ToString());
+            await shardRef.UpdateAsync("count", FieldValue.Increment(1));
+        }
+        // [END fs_increment_counter]
+
+        // [START fs_get_count]
+        /// <summary>
+        /// Get total count across all shards.
+        /// </summary>
+        /// <param name="docRef">The document reference <see cref="DocumentReference"/></param>
+        /// <returns>The <see cref="int"/></returns>
+        private static async Task<int> GetCountAsync(DocumentReference docRef)
+        {
+            var snapshotList = await docRef.Collection("shards").GetSnapshotAsync();
+            return snapshotList.Sum(shard => shard.GetValue<int>("count"));
+        }
+        // [END fs_get_count]
+
+        public static void Main(string[] args)
+        {
+            if (args.Length < 2)
+            {
+                Console.Write(Usage);
+                return;
+            }
+            string command = args[0].ToLower();
+            string projectId = string.Join(" ",
+                new ArraySegment<string>(args, 1, args.Length - 1));
+            switch (command)
+            {
+                case "distributed-counter":
+                    FirestoreDb db = FirestoreDb.Create(projectId);
+                    const int numberOfShards = 5;
+                    var docRef = db.Collection("counter_samples").Document("DCounter");
+
+                    CreateCounterAsync(docRef, numberOfShards).Wait();
+                    Console.WriteLine("Distributed counter created.");
+
+                    IncrementCounterAsync(docRef, numberOfShards).Wait();
```

Comment:
nit: In this wait and the above, consider using Task.Run().WaitWithUnwrappedExceptions similar to what you are doing a couple of lines down.